### PR TITLE
@nrwl/bazel schematics initial release

### PR DESCRIPTION
**Pull request by mrmeku**
Supported Commands:

Add new app
node_modules/.bin/ng generate app [name] --collection=@nrwl/bazel

Example: node_modules/.bin/ng generate app test

Add new module
node_modules/.bin/ng generate module [name]

Example: node_modules/.bin/ng generate component friends --directory=myDir
* must manually import the component's Bazel rule in the consuming Bazel rule

Add new lib
node_modules/.bin/ng generate lib  [name]

Example: node_modules/.bin/ng generate lib  mylib
* must manually import the lib's Bazel rule in the consuming Bazel rule

Run dev server
ibazel run apps/[app specific path]]/src:devserver (anything between apps/**/src points to a specific app)

Example: ibazel run apps/my-dir/my-app/src:devserver

Run prod server
bazel run apps/[app specific path]]/src:prodserver (anything between apps/**/src points to a specific app)

Example: bazel run apps/my-dir/my-app/src:prodserver

Run unit tests
ibazel test //libs/mylib/src:test
* currently works for libs

### CODE REVIEW

File: e2e/schematics/bazel-application.test.ts [hunk 1]
```diff
@@ -1,15 +1,165 @@
...
+//   runCommand,
+//   updateFile,
+//   cleanup
+// } from '../utils';
+
+// function itShould(testDescription, test) {
+//   return it(`should ${testDescription}`, test, 100000);
+// }
+
+// describe('Nrwl Workspace (Bazel)', () => {
```
Comment (vsavkin) [inline e2e/schematics/bazel-application.test.ts]: CI fails cause just requires you to have at least one test in this files.

File: e2e/schematics/application.test.ts [hunk 2]
```diff
@@ -37,7 +52,9 @@ describe('Nrwl Workspace', () => {
     () => {
       newProject();
       newApp('myApp --directory=myDir --routing');
-      newLib('myLib --directory=myDir --routing --lazy --parentModule=apps/my-dir/my-app/src/app/app.module.ts');
+      newLib(
```
Comment (vsavkin) [inline e2e/schematics/application.test.ts]: have you changed the formatting options for this project?

File: e2e/utils.ts [hunk 3]
```diff
@@ -83,12 +84,26 @@ export function runCLI(
   }
 }
 
-export function newApp(name: string): string {
-  return runCLI(`generate app ${name}`);
+export function newApp(name: string, schematics?: string): string {
```
Comment (vsavkin) [inline e2e/utils.ts]: None of these helpers functions should differ for bazel and non-bazel setups. The user should never worry about passing the right collection, and instead should just call:

```
ng g app myapp
ng g component mycomponent
```

To make it work, update `.angular-cli.json` to set the default collection to `@nrwl/bazel`.

File: packages/bazel/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+Bazel Commands
+
+## Add new app
+node_modules/.bin/ng generate app [name] --collection=@nrwl/bazel
```
Comment (vsavkin) [inline packages/bazel/README.md]: Remove --collection=@nrwl/bazel

File: e2e/schematics/application.test.ts [hunk 5]
```diff
@@ -1,6 +1,21 @@
...
+  newLib,
+  newProject,
+  runCLI,
+  updateFile,
+  cleanup
+} from '../utils';
 
 describe('Nrwl Workspace', () => {
+  beforeAll(() => {
+    cleanup();
```
Comment (vsavkin) [inline e2e/schematics/application.test.ts]: why is this required?

`newProject()` will always do cleanup, right?  And every test invokes it.

**Review by vsavkin — DISMISSED** [review]
Left some comments.

**Comment by vsavkin** [comment]
I have a quick question. 

Do we have to generate an ngmodule per component? Can we have one per lib (or however many folks want)? As far as I understand, the only pro is smaller compilation units.

The cons are:

- it's a serious change of the CLI behavior.
- It forces folks to be more aware of Bazel (every time I add a component, instead of every time I add a lib)
- It won't be necessary once angular compiles components individually (with ngIvy)

What do you think?

**Comment by kylecordes** [comment]
@vsavkin  Here's an attempt to answer your question outside:

I think many developers using these tools (including me and everyone here) primarily want to see Bazel as a different way of compiling the same code. The ideal "diff" to add Bazel to a project would be automatic creation of a BUILD file per compilation unit, and little other change. This would imply not generating any different NgModules or other structural changes, but rather just adding a BUILD file per NgModule.

This is compatible also with the broader idea of using NgModules as units of modularity in the more general sense; not just a syntactical tax imposed by Angular, but rather a semantic unit of intended cohesion. Not too big, not too small.

File: e2e/schematics/application.test.ts (hunk 2 above)
Comment (mrmeku) [reply e2e/schematics/application.test.ts]: I didn't make any changes. I think this file just might have been formatted with a previous release of prettier

File: e2e/schematics/application.test.ts (hunk 5 above)
Comment (mrmeku) [reply e2e/schematics/application.test.ts]: Its not. I'll remove all of the occurances

**Comment by mrmeku** [comment]
@vsavkin @kylecordes - Currently the ng_module build rules enforces that every component must have an associated NgModule annotated class within the same directory. This is a limitation imposed by the build rule itself rather than a preference for more incremental builds.

**Comment by kylecordes** [comment]
@mrmeku  I have not found that to be the case; ng_module seems happy with component source code located in a subdirectory of where the BUILD is located, as long as the paths in the BUILD are updated appropriately. I'll put a screenshot below, showing what I did, making a slight edit to Alex Eagle's Angular-Bazel example. It appears to be merely enforcing the same rules as `ngc` does behind the scenes, and that for a unit of compilation every component has to be in a module.

(One caveat: I couldn't get Sass to work, in 5 minutes of trying. I don't know if this is a major limitation I've uncovered or just some bit I was doing wrong.)

screen shot 2018-03-12 at 6 47 36 pm

**Comment by mrmeku** [comment]
@kylecordes I'm so happy to be wrong about that! That's actually really exciting to me. So theoretically, a bazel related schematic for component generation so not generate a build file by default, and should instead just amend the generated .ts files to an existing ng_module rule in a parent directory. I can try to work on doing that this week. Thanks for your input!

File: e2e/schematics/application.test.ts (hunk 2 above)
Comment (vsavkin) [reply e2e/schematics/application.test.ts]: Can you rebase? All the formatting issues should go away

File: packages/schematics/src/collection/app/index.ts [hunk 6]
```diff
@@ -247,20 +271,23 @@ export default function(schema: Schema): Rule {
...
+      updateComponentTemplate(options), addBootstrap(options.fullPath),
+      addNxModule(options.fullPath), addAppToAngularCliJson(options),
       options.routing ? addRouterRootConfiguration(options.fullPath) : noop()
     ]);
   });
 }
 
 function normalizeOptions(options: Schema): NormalizedSchema {
   const name = toFileName(options.name);
+<<<<<<< HEAD
```
Comment (vsavkin) [inline packages/schematics/src/collection/app/index.ts]: this should not be here

File: packages/schematics/src/collection/lib/index.ts [hunk 7]
```diff
@@ -232,49 +225,70 @@ function addChildren(schema: NormalizedSchema): Rule {
   };
 }
 
-export default function(schema: Schema): Rule {
-  return wrapIntoFormat(() => {
-    const options = normalizeOptions(schema);
-    const moduleFileName = `${toFileName(options.name)}.module`;
-    const modulePath = `${options.fullPath}/${moduleFileName}.ts`;
-    const indexFile = `libs/${toFileName(options.fullName)}/index.ts`;
+export interface LibSchema {
```
Comment (vsavkin) [inline packages/schematics/src/collection/lib/index.ts]: individual schematics should not export tokens. if you think the functionality should be reused, moved into some shared file

File: packages/bazel/src/collection/lib/index.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import {branchAndMerge, chain, mergeWith, Rule,} from '@angular-devkit/schematics';
+
+import {validateLibSchema} from '../../../../schematics/src/collection/lib';
```
Comment (vsavkin) [inline packages/bazel/src/collection/lib/index.ts]: even though it happened to work right now, this won't work in a prod env. bazel and schematics are too separate packages.

So you will have to import it using `@nrwl/schematics`.

I think it might be easer to just duplicate this functionality

File: package.json [hunk 9]
```diff
@@ -5,16 +5,17 @@
...
     "e2e": "./scripts/e2e.sh",
     "format": "precise-commits --whitelist=\"{packages,e2e}/**/*.ts\"",
     "linknpm": "./scripts/link.sh",
     "package": "./scripts/package.sh",
     "release": "./scripts/release.sh",
     "copy": "./scripts/copy.sh",
     "test:schematics": "yarn linknpm && ./scripts/test_schematics.sh",
-    "test:nx": "yarn linknpm && ./scripts/test_nx.sh",
-    "test": "yarn linknpm && ./scripts/test_nx.sh && ./scripts/test_schematics.sh",
+    "test:nx": "./scripts/link-spec.sh && ./scripts/test_nx.sh",
```
Comment (vsavkin) [inline package.json]: I'm not sure why `link-spec` is required given that test:schematics still invokes `linknpm`.

File: package.json [hunk 10]
```diff
@@ -5,16 +5,17 @@
   "main": "index.js",
   "private": true,
   "scripts": {
-    "build": "./scripts/build.sh",
+    "build": "ng-packagr -p packages/nx/ng-package.json && ./scripts/build.sh",
```
Comment (vsavkin) [inline package.json]: previously we had the build and package commands. The first one would compile things and copy statics, the second one would make them ready to package.

The build command has changed to do the latter. Having the two commands in the way they are written now doesn't feel right. We either need to move the packaging functionality from build to package, or merge the two commands.

**Review by vsavkin — DISMISSED** [review]

File: package.json (hunk 9 above)
Comment (mrmeku) [reply package.json]: Its now been removed, I combined the two scripts. The issue we had is that ngpackagr takes an entry point to determine which files to package. The entry point did not point to any spec files so we needed a different building mechanism to transpile our spec.ts files
